uk.co.compucorp.civicrm.hrcontactactionsmenu/CRM/HRContactActionsMenu/Page/UserAccount.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+use CRM_HRContactActionsMenu_Helper_Contact as ContactHelper;
+use CRM_HRCore_CMSData_UserAccountFactory as UserAccountFactory;
+
+/**
+ * Class CRM_HRContactActionsMenu_Page_UserAccount
+ */
+class CRM_HRContactActionsMenu_Page_UserAccount {
+
+  /**
+   * Function to delete the CMS user account of the
+   * contact.
+   */
+  public static function delete() {
+    $contactID = CRM_Utils_Array::value('cid', $_GET);
+    $contactInfo = ContactHelper::getUserInformation($contactID);
+    $userAccount = UserAccountFactory::create();
+    $userAccount->cancel($contactInfo);
+
+    CRM_Core_Session::setStatus(ts('User account has been deleted'), 'Success', 'success');
+
+    $url = CRM_Utils_System::url('civicrm/contact/view', "reset=1&cid=$contactID");
+    CRM_Utils_System::redirect($url);
+  }
+
+  /**
+   * Function to disable the CMS user account of the
+   * contact.
+   */
+  public static function disable() {
+    $contactID = CRM_Utils_Array::value('cid', $_GET);
+    $contactInfo = ContactHelper::getUserInformation($contactID);
+    $userAccount = UserAccountFactory::create();
+    $userAccount->disable($contactInfo);
+
+    CRM_Core_Session::setStatus(ts('User account has been disabled'), 'Success', 'success');
+
+    $url = CRM_Utils_System::url('civicrm/contact/view', "reset=1&cid=$contactID");
+    CRM_Utils_System::redirect($url);
+  }
+
+  /**
+   * Function to enable the CMS user account of the
+   * contact.
+   */
+  public static function enable() {
+    $contactID = CRM_Utils_Array::value('cid', $_GET);
+    $contactInfo = ContactHelper::getUserInformation($contactID);
+    $userAccount = UserAccountFactory::create();
```

Comment:
These 3 lines are repeated in each of the methods. How about extracting them to a separate method?

Reply:
Its not straightforward to extract, reason is the $contactInfo is needed to perform necessary action on the User object and $contactID is needed to get the contact Info and also for the redirection.
Any Ideas?

Reply:
I see. How about extract the only first two lines then?

Reply:
Issue is I would still need the first line
`$contactID = CRM_Utils_Array::value('cid', $_GET);` so I can pass the contact ID to the self::redirectToContactSummaryPage function.
Except I have to duplicate the `$contactID = CRM_Utils_Array::value('cid', $_GET);` in the self::redirectToContactSummaryPage function function and then in this function we want to extract the first two lines to which doesn't seem to be necessary to me.

Reply:
How about this:

```php
private static function doUserAction($action, $successMessage) {
    $contactID = CRM_Utils_Array::value('cid', $_GET);
    $contactInfo = ContactHelper::getUserInformation($contactID);
    $userAccount = UserAccountFactory::create();
    $userAccount->$action($contactInfo);
    CRM_Core_Session::setStatus(ts($successMessage), 'Success', 'success');

    $url = CRM_Utils_System::url('civicrm/contact/view', "reset=1&cid=$contactID");
    CRM_Utils_System::redirect($url);
}
```

And then each callback method would be like:

```php
public function enable() {
  self::doUserAction('enable', 'User account has been enabled');
}

// and so on
```